src/bootstrap/src/core/build_steps/check.rs

```diff
@@ -410,18 +434,34 @@ macro_rules! tool_check_step {
 fn run_tool_check_step(
     builder: &Builder<'_>,
     target: TargetSelection,
-    step_type_name: &str,
     path: &str,
+    mode: Mode,
+    allow_features: &str,
 ) {
     let display_name = path.rsplit('/').next().unwrap();
-    let compiler = builder.compiler(builder.top_stage, builder.config.build);
 
-    builder.ensure(Rustc::new(target, builder));
+    let host = builder.config.build;
+    let compiler;
+
+    match mode {
+        Mode::ToolBootstrap => {
+            compiler = builder.compiler(0, host);
+        }
+        Mode::ToolStd => {
+            compiler = builder.compiler(0, host);
+            builder.ensure(Std::new(target).with_compiler(compiler));
```

Comment:
We also have similar macros for building tools. Why can these work without `with_compiler`?

Reply:
The `compile::Std` step takes an explicit compiler argument, so the equivalent of `with_compiler` is built into the constructor.

The `check::Std` step does not, and is hardcoded to use the `top_stage` compiler unless overridden.

Reply:
I see. Yeah, that's a common situation in bootstrap, when you touch something, it's often immediately time for refactoring of the previous state :D